x/mongo/driver/topology/connection_test.go

```diff
@@ -252,6 +252,140 @@ func TestConnection(t *testing.T) {
+						// Ensure the initial connection dial can be timed out and the connection propagates the error
+						// from the dialer in this case.
+
+						connOpts := []ConnectionOption{
+							WithDialer(func(Dialer) Dialer {
+								return DialerFunc(func(ctx context.Context, _, _ string) (net.Conn, error) {
+									<-ctx.Done()
+									return nil, ctx.Err()
+								})
+							}),
+							WithConnectTimeout(func(time.Duration) time.Duration {
+								return tc.connectTimeout
+							}),
+						}
+						conn, err := newConnection("", connOpts...)
+						assert.Nil(t, err, "newConnection error: %v", err)
+
+						ctx, cancel := context.WithTimeout(context.Background(), tc.contextTimeout)
+						defer cancel()
+						var connectErr error
+						callback := func() {
+							conn.connect(ctx)
+							connectErr = conn.wait()
+						}
+						assert.Soon(t, callback, tc.maxConnectTime)
+
+						ce, ok := connectErr.(ConnectionError)
+						assert.True(t, ok, "expected error %v to be of type %T", connectErr, ConnectionError{})
+						assert.Equal(t, context.DeadlineExceeded, ce.Unwrap(), "expected wrapped error to be %v, got %v",
+							context.DeadlineExceeded, ce.Unwrap())
+					})
+					t.Run("timeout applied to TLS handshake: "+tc.name, func(t *testing.T) {
+						// Ensure the TLS handshake can be timed out and the connection propagates the error from the
+						// tlsConn in this case.
+
+						var hangingTLSConnectionSource tlsConnectionSourceFn = func(nc net.Conn, cfg *tls.Config) tlsConn {
+							tlsConn := tls.Client(nc, cfg)
+							return newHangingTLSConn(tlsConn, tc.maxConnectTime)
+						}
+
+						connOpts := []ConnectionOption{
+							WithConnectTimeout(func(time.Duration) time.Duration {
+								return tc.connectTimeout
+							}),
+							WithDialer(func(Dialer) Dialer {
+								return DialerFunc(func(context.Context, string, string) (net.Conn, error) {
+									return &net.TCPConn{}, nil
+								})
+							}),
+							WithTLSConfig(func(*tls.Config) *tls.Config {
+								return &tls.Config{}
+							}),
+							withTLSConnectionSource(func(tlsConnectionSource) tlsConnectionSource {
+								return hangingTLSConnectionSource
+							}),
+						}
+						conn, err := newConnection("", connOpts...)
+						assert.Nil(t, err, "newConnection error: %v", err)
+
+						ctx, cancel := context.WithTimeout(context.Background(), tc.contextTimeout)
+						defer cancel()
+						var connectErr error
+						callback := func() {
+							conn.connect(ctx)
+							connectErr = conn.wait()
+						}
+						assert.Soon(t, callback, tc.maxConnectTime)
+
+						ce, ok := connectErr.(ConnectionError)
+						assert.True(t, ok, "expected error %v to be of type %T", connectErr, ConnectionError{})
+						assert.Equal(t, context.DeadlineExceeded, ce.Unwrap(), "expected wrapped error to be %v, got %v",
+							context.DeadlineExceeded, ce.Unwrap())
+					})
+					t.Run("timeout is not applied to handshaker: "+tc.name, func(t *testing.T) {
```

Comment:
This test doesn't actually use any material from the testcases. Is that on purpose? If so, this can be run once.

Reply:
Updated the test to use `tc.connectTimeout`.